packages/material-ui-system/src/index.d.ts

```diff
@@ -161,7 +161,7 @@ export interface StyleOptions<PropKey, Theme extends object> {
 }
 export function style<PropKey extends string, Theme extends object>(
   options: StyleOptions<PropKey, Theme>,
-): StyleFunction<{ [K in PropKey]: unknown } & { theme: Theme }>;
+): StyleFunction<{ [K in PropKey]?: unknown }>;
```

Comment:
There is a catch. This is wrong. Style function actually requires theme. So when you use it like this:
```js
assert.deepEqual(border({
  border: 1,
}),{
  border: '1px silid',
});
```
You will get:
`Error: Warning: @material-ui/system: you are calling a style function without a theme value.`

However this typing is also inferred as Props for the enhanced component, which make the component require the theme, even thought its provided via context or `defaultTheme`.

This need to be solved differently. Need to look into it.

Reply:
I'm not following. If it is required then the types should reflect that. Could you write a simple usage of `style` that currently works at runtime but is rejected by the type-checker?

Reply:
@eps1lon 
https://codesandbox.io/s/stoic-yonath-3ukmk

So what my comment means is: the `theme is required`-issue is not the typing for the `StyleFunction`, which is returned from `style`, but the `styled`-enhancer/hoc.

The problem is, that making theme not required is not 100% correct, as it might be required. In my codesandbox example it would be the case when I remove the ThemeProvider.

However I see no way in typescript to type it correctly in case of context. But the issue with theme always required, when using `style`-created StyleFunction makes it kind of absurd.

Notably it seems like the typings for the built-in style functions like `palette` do use the same style function:
https://github.com/mui-org/material-ui/blob/02671c05c0eefef43fc126845da4b2a04052c208/packages/material-ui-system/src/palette.js#L1-L20

And therefore the created style-function do require theme property.
However they work with `styled` enhancer, because the theme property is not exposed in typing by using another interface:
https://github.com/mui-org/material-ui/blob/4e9fff3d8a22a7cf6d202a743ec6bc02f002a98a/packages/material-ui-system/src/index.d.ts#L87

So now what can be done:
1.
Instead of removing theme key here:
```ts
StyleFunction<{ [K in PropKey]: unknown } & { theme: Theme }
````
Make it "nullable":
```ts
StyleFunction<{ [K in PropKey]: unknown } & { theme?: Theme }
````
2.
Try change typing for the enhancer/hoc created by `styleD` make the theme property nullable, leave StyleFunction as is (theme required).
3. 
Leave style as is and make it absurd to use.

1. Has the advantage of being conform to how you expose buil-it style functions (hiding away the fact, that theme is needed). This is however a runtime error, like it would be one with the built-ins.

2. Would make at least direct StyleFunction usage a typecheck error, when you do not provide theme. However you might still omit theme when one is required and get a runtime error.

Reply:
@eps1lon 
https://codesandbox.io/s/stoic-yonath-3ukmk

So what my comment means is: the `theme is required`-issue is not due to the typing for the `style` typing, but due to the `styleD`-enhancer/hoc.

The problem is: making theme not required is not always correct, as it might be required. In my codesandbox example it would be the case when I remove the ThemeProvider.

However I see no way in typescript to type it correctly in case of context. AFAIK its opaque by purpose. But the issue with theme always being required, when using `style`-created StyleFunction, makes its usage kind of absurd.

Notably it seems like the the built-in style functions like `palette` do use the same style function:
https://github.com/mui-org/material-ui/blob/02671c05c0eefef43fc126845da4b2a04052c208/packages/material-ui-system/src/palette.js#L1-L20

And therefore the created style-function do require theme property.
However they work with `styled` enhancer, because the theme property is not exposed in typing by using another interface:
https://github.com/mui-org/material-ui/blob/4e9fff3d8a22a7cf6d202a743ec6bc02f002a98a/packages/material-ui-system/src/index.d.ts#L87

So now what can be done:

A
Instead of removing theme key here:
```ts
StyleFunction<{ [K in PropKey]: unknown } & { theme: Theme }
````
Make it "nullable":
```ts
StyleFunction<{ [K in PropKey]: unknown } & { theme?: Theme }
````

B
Try change typing for the enhancer/hoc created by `styleD` make the theme property nullable, leave StyleFunction as is (theme required).

C
Leave style as is and make it absurd to use.

A. Has the advantage of being conform to how you expose buil-it style functions (hiding away the fact, that theme is needed). This is however a runtime error, like it would be one with the built-ins.

B. Would make at least direct StyleFunction usage a typecheck error, when you do not provide theme. However you might still omit theme when one is required and get a runtime error on components.

C. sucks :D

Reply:
So you have the code. Add it as a test and make it runnable if this is you goal. If you can't do it I might take a look. But I either dive in fully or give you guidance about project layout. I won't have the time to explain every technical detail.

Reply:
@eps1lon 

typing for style is totally untested and has tons of errors. My example is not all I need, however as I explained adding a test for each single type error is impossible because the moment you add the first type test for anything related to style all the other issues suddenly fail test. I have to simply fit the typing for style and make it actually tested and not fiddle with parts of its issues.

Reply:
I was just asking you to keep changes as atomic as possible but if that's not viable don't do it. Instead you're complaining again and again how bad things are and how broken things are and how you did better. Be a bit more constructive please. 

> asked whether you really want this discord by design for some reason.

Why would I want broken typings? If I gave you the impression that my typings are gospel and cannot be improved I apologize. In fact I'm very much looking for more TypeScript contributors. But so far you didn't contribute much apart from telling me I'm supposedly wrong. I would appreciate if you just push your tests and changes to this PR so that we can get to work.

Reply:
@eps1lon I apologize, that you feel like you are being attacked, if you feel this way a while already this explains why we don't get further because such impression may be reflected in constructiveness of own responses, without noticing it. Do make it clear, I am here not to blame. Every single word I wrote took me time to write to give you feedback about the issue and reaction to you responses. If it was required to point out, that you can not simply add a single test because the tests were missing and the typing all wrong until now, adding such a test would bring it all to daylight and require it fixed too. I do not know how differently I could explain it to you and if you felt blamed for something, as said I apologize. However you still did not answer my question and we are continuing to waste time.

Reply:
Could you just add the tests that illustrate the issue instead of focusing on the implementation?

Reply:
I was talking about Typings, not implementation. Even the built-in example showing the js file was related to typings only. 

I still do not have the required answers. I will just push all the tests/changes I did as atomic as possible and we can maybe understand each other better when looking at it :)

Reply:
I can't give you every answer before seeing tests or how your changes look. It just isn't efficient to fully look into possible implementations and get all the context. At that point I could just do it myself. This contributor - maintainer relationship just doesn't work if you need all answers before implementation. Especially for types.  So next time I don't get it just open the PR so that I can look at it from another angle.